client/my-sites/plans-v2/details.tsx

```diff
@@ -2,20 +2,155 @@
  * External dependencies
  */
 import React from 'react';
+import { useSelector } from 'react-redux';
+import page from 'page';
+import { useTranslate } from 'i18n-calypso';
 
 /**
  * Internal dependencies
  */
-import { DetailsPageProps } from './types';
-import { TERM_ANNUALLY } from 'lib/plans/constants';
+import {
+	slugToSelectorProduct,
+	durationToString,
+	durationToText,
+	productButtonLabel,
+	getProductPrices,
+} from './utils';
+import {
+	PRODUCTS_WITH_OPTIONS,
+	OPTIONS_JETPACK_SECURITY,
+	OPTIONS_JETPACK_SECURITY_MONTHLY,
+} from './constants';
+import FormattedHeader from 'components/formatted-header';
+import HeaderCake from 'components/header-cake';
+import Main from 'components/main';
+import JetpackBundleCard from 'components/jetpack/card/jetpack-bundle-card';
+import JetpackProductCard from 'components/jetpack/card/jetpack-product-card';
+import { getCurrentUserCurrencyCode } from 'state/current-user/selectors';
+import { getAvailableProductsList } from 'state/products-list/selectors/get-available-products-list';
+import { isProductsListFetching } from 'state/products-list/selectors/is-products-list-fetching';
+import { getProductCost } from 'state/products-list/selectors/get-product-cost';
+import { getSelectedSiteSlug } from 'state/ui/selectors';
+
+/**
+ * Type dependencies
+ */
+import type { DetailsPageProps, PurchaseCallback, SelectorProduct } from './types';
+
+import './style.scss';
+
+type Cost = {
+	originalPrice: number;
+	discountedPrice?: number;
+};
+
+const ProductComponent = ( {
+	productSlug,
+	onClick,
+	currencyCode,
+}: {
+	productSlug: string;
+	onClick: PurchaseCallback;
+	currencyCode: string;
+} ) => {
+	const availableProducts = useSelector( ( state ) => getAvailableProductsList( state ) );
+	const price = useSelector( ( state ) => getProductCost( state, productSlug ) ) || 0;
+	const product = slugToSelectorProduct( productSlug );
+	if ( ! product ) {
+		// Exit if invalid product.
+		return null;
```

Comment:
Wouldn't it be better if we redirect users back to the Selector page in this case? Maybe it's something we can verify at the controller level.

Reply:
The code does redirect users [farther down on this file](https://github.com/Automattic/wp-calypso/pull/44849/files#diff-6a81e65088668c4d58d11412d0ed9f25R107-R118). This is the `ProductComponent`, which should return `null` if the data is invalid as there's nothing to display.

As for the controller level, I looked into that, but the issue I found is that we don't have simple access to the state there without things getting messy. We need the state to get `siteSlug` for a proper redirect.

Reply:
One thing I will add is that I think we can probably unify this pattern into a single component, as right now this is needed repeatedly to pull in price data and such from the state. I'd want to do that in a separate PR however.

Reply:
Sure, sounds like that would deserve its own PR.